FIAT/morley.py

```diff
@@ -5,60 +5,56 @@
 # SPDX-License-Identifier:    LGPL-3.0-or-later
 
 from FIAT import finite_element, polynomial_set, dual_set, functional
-from FIAT.reference_element import TRIANGLE, ufc_simplex
+from FIAT.reference_element import TETRAHEDRON, TRIANGLE
+from FIAT.quadrature import FacetQuadratureRule
 from FIAT.quadrature_schemes import create_quadrature
 import numpy
 
 
 class MorleyDualSet(dual_set.DualSet):
-    """The dual basis for Lagrange elements.  This class works for
-    simplices of any dimension.  Nodes are point evaluation at
-    equispaced points."""
+    """The dual basis for Morley elements.  This class works for
+    simplices of any dimension.  Nodes are average on codim=2 entities
+    and average of normal derivative on codim=1 entities."""
 
-    def __init__(self, ref_el):
-        entity_ids = {}
-        nodes = []
-        cur = 0
-
-        # make nodes by getting points
-        # need to do this dimension-by-dimension, facet-by-facet
+    def __init__(self, ref_el, degree):
         top = ref_el.get_topology()
-        verts = ref_el.get_vertices()
-        if ref_el.get_shape() != TRIANGLE:
-            raise ValueError("Morley only defined on triangles")
-
-        # vertex point evaluations
-
-        entity_ids[0] = {}
-        for v in sorted(top[0]):
-            nodes.append(functional.PointEvaluation(ref_el, verts[v]))
-
-            entity_ids[0][v] = [cur]
-            cur += 1
-
-        # edge dof -- average of normal derivative at each edge
-        rline = ufc_simplex(1)
-        degree = 2
-        Q = create_quadrature(rline, degree-1)
-        qwts = Q.get_weights()
-        scale = numpy.ones(qwts.shape)
-
-        entity_ids[1] = {}
-        for e in sorted(top[1]):
-            n = functional.IntegralMomentOfNormalDerivative(ref_el, e, Q, scale)
-            nodes.append(n)
-            entity_ids[1][e] = [cur]
-            cur += 1
-
-        entity_ids[2] = {0: []}
+        sd = ref_el.get_spatial_dimension()
+        entity_ids = {dim: {entity: [] for entity in top[dim]} for dim in top}
+        nodes = []
+        for codim in (2, 1):
+            dim = sd - codim
+            facet = ref_el.construct_subelement(dim)
+            Q_ref = create_quadrature(facet, degree+codim-2)
+            scale = numpy.ones(Q_ref.get_weights().shape)
+
+            for entity in sorted(top[dim]):
+                cur = len(nodes)
+                if dim == sd-1:
+                    # codim=1 dof -- average of normal derivative at each facet
+                    ell = functional.IntegralMomentOfNormalDerivative(ref_el, entity, Q_ref, scale)
+                elif dim == 0:
```

Comment:
Better: Make `IntegralMoment` DTRT and reduce to point evaluation on 0-dimensional facets.  This would match the maths and the paper.